presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -134,94 +194,377 @@ public boolean needsInput()
-                    if (!currentProbePositionProducedRow) {
-                        currentProbePositionProducedRow = true;
-                        if (!outerJoinCurrentPosition(lookupSource)) {
-                            break;
-                        }
-                    }
-                }
-                currentProbePositionProducedRow = false;
-                if (!advanceProbePosition(lookupSource)) {
-                    break;
                 }
-                statisticsCounter.recordProbe(joinSourcePositions);
-                joinSourcePositions = 0;
             }
+            currentProbePositionProducedRow = false;
+            if (!advanceProbePosition(lookupSource)) {
+                break;
+            }
+            statisticsCounter.recordProbe(joinSourcePositions);
+            joinSourcePositions = 0;
         }
+    }
+
+    private void restoreProbe(Page probePage, long joinPosition, boolean currentProbePositionProducedRow, int joinSourcePositions, SpillInfoSnapshot spillInfoSnapshot)
+    {
+        verify(probe == null);
+
+        addInput(probePage, spillInfoSnapshot);
+        verify(probe.advanceNextPosition());
+        this.joinPosition = joinPosition;
+        this.currentProbePositionProducedRow = currentProbePositionProducedRow;
+        this.joinSourcePositions = joinSourcePositions;
+    }
+
+    private Page pageTail(Page currentPage, int startAtPosition)
+    {
+        verify(currentPage.getPositionCount() - startAtPosition >= 0);
+
+        int[] retainedPositions = new int[currentPage.getPositionCount() - startAtPosition];
+        Arrays.setAll(retainedPositions, i -> startAtPosition + i);
+        return mask(currentPage, retainedPositions);
+    }
 
-        // only flush full pages unless we are done
-        if (pageBuilder.isFull() || (finishing && !pageBuilder.isEmpty() && probe == null)) {
+    public static class SavedRow
+    {
+        /**
+         * A page with exactly one {@link Page#getPositionCount}, representing saved row.
+         */
+        public final Page row;
+
+        /**
+         * A snapshot of {@link LookupJoinOperator#joinPosition} "de-partitioned", i.e. {@link LookupJoinOperator#joinPosition} is a join position
+         * with respect to (potentially) partitioned lookup source, while this value is a join position with respect to containing partition.
+         */
+        public final long joinPositionWithinPartition;
+
+        /**
+         * A snapshot of {@link LookupJoinOperator#currentProbePositionProducedRow}
+         */
+        public final boolean currentProbePositionProducedRow;
+
+        /**
+         * A snapshot of {@link LookupJoinOperator#joinSourcePositions}
+         */
+        public final int joinSourcePositions;
+
+        public SavedRow(Page page, int position, long joinPositionWithinPartition, boolean currentProbePositionProducedRow, int joinSourcePositions)
+        {
+            this.row = mask(page, new int[] {position});
```

Comment:
This will retain the entire page.  I believe you want to build a new Page by calling `getSingleValueBlock` on each block.

Reply:
no, because i call `compact()`?